### DISCUSSION TITLE
Move shared lambda code into a layer

### DISCUSSION THREAD

**Discussion by aphex** [architect/architect#1259] [Feature Requests]
Moving shared code into a Lambda layer will help keep each lambda smaller and possibly reduce deploy times for projects with a very large amount of cloud functions. Likely some other benefits I am not thinking of.

https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html

**Comment by ryanblock** [comment]
It's an idea we've heard before. Unfortunately, I think your premises are wrong:

> [it] will help keep each lambda smaller

Not really. See: https://docs.aws.amazon.com/lambda/latest/dg/invocation-layers.html (specifically `The total unzipped size of the function and all layers cannot exceed the unzipped deployment package size quota of 250 MB.`). In this case, adding a layer simply moves code from one place to another.

> possibly reduce deploy times

Publishing a new layer with each code change will do exactly the opposite, increasing deploy times (not to mention significantly increasing the complexity of each deployment).

For these reasons (and others) we've opted not to utilize layers in this fashion. However, I could see per-deployment layer publishing making an interesting [Architect plugin](https://arc.codes/docs/en/guides/extend/plugins) for folks who have esoteric reasons to do so.

**Comment by aphex** [comment]
@ryanblock I think I see where you're going with the first point. Technically the total runtime lambda would not be smaller as it still loads the same amount of code. However from what I understand lambda is capped at 50mb, right? I thought when using layers it is then extended to 250mb. This would then allow a dev to have more room. However lets be honest you really don't want to be near anything like those sizes. 

Another benefit, from what I understand please correct if I am wrong, but layers are "always on" and mounted in the `opt` directory. So during initialization this code would not need to be extracted. Also is there any testing on execution time around this? Seeing that the layer is always there, do we save MS on lambda runs, which could translate into $$$?

On your second point; wouldn't the layer just be deployed once and then each function would only deploy its "non-shared" code? So at build/deploy time there would be some savings? Maybe this is the per-deployment plugin you mentioned, but I would just assume that's how it would work OOTB, especially on larger projects with a lot of functions.

Finally wouldn't we also see a benefit when updating only shared code? One would just need to update the layer and all lambdas would then be updated?

I am thinking the benefits here would be rather small if you have only a few functions or maybe your shared code is minimal, however for a larger project this could be a bigger deal. Maybe it is simply not worth the Dev effort right now as the recommended use case for ARC/shared code doesn't fall into these larger scale projects?

**Comment by ryanblock** [comment]
> However from what I understand lambda is capped at 50mb, right? I thought when using layers it is then extended to 250mb. This would then allow a dev to have more room.

Again, sorry, no, that is not the case. See: https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html
```
50 MB (zipped, for direct upload)
250 MB (unzipped)
This quota applies to all the files you upload, including layers and custom runtimes.
3 MB (console editor)
```

> layers are "always on" and mounted in the opt directory.

I'm not sure where you got that impression, but I don't believe that's the case, and I've never heard anything of the sort. Some reading:

- https://docs.aws.amazon.com/lambda/latest/operatorguide/execution-environments.html
- https://docs.aws.amazon.com/lambda/latest/dg/best-practices.html

Again, see the note above when referring to the file quotas applying to layers. I think it's pretty clear that layers are treated the same as code in terms of the Lambda init lifecycle.

> wouldn't the layer just be deployed once and then each function would only deploy its "non-shared" code? So at build/deploy time there would be some savings?

It's theoretically feasible that there might be some savings, but that assumes publishing code to Lambda is slower than publishing a new layer via CloudFormation. I wouldn't take that bet – but if it was faster, I think it would be negligible.

> Maybe it is simply not worth the Dev effort right now as the recommended use case for ARC/shared code doesn't fall into these larger scale projects?

Not sure where that came from. Architect is intended for large-scale production use, and has been in production in very sizeable projects and organizations for years.

But good news: if you really want this feature, Architect should support it as a plugin without much issue. Some links that might be helpful:

- https://arc.codes/docs/en/guides/extend/plugins
- https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html#invocation-layers-cloudformation
- https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-layers.html

**Reply by aphex** [reply]
These are all great links and references and honestly I don't know if I really want this feature :) from the sounds of it things I am reading are getting mixed up. 

I had read the 50MB lambda and 250MB with layers, if that's not the case great! no need to worry about that.  This is where that came from.
https://www.n-k.de/2019/02/aws-lambda-layers-pros-and-cons.html#:~:text=Just%20a%20layer.-,Pros,-On%20the%20pro
This person also notes on initialization time in there as well. I had read that in a couple other places as well. 

If this has been tested and is simply not the case, then this is an easy answer. Downside to the internet is opinions sometimes become facts. Goal of this is to simply help Architect get better. Very possible this is not a path to that.

So out of curiosity why do you think these things exist? Simply to share more intense code, or larger libraries from a single place, like a registry? like chromium with puppeteer or playwright? I guess the thought here is if it is not more efficient to put code into layers that is shared, why does it exist? 🤷‍♂️

**Comment by ryanblock** [comment]
Yeah, I'm sorry to say that the person who authored that link does not seem to know what they're talking about. And who knows, he may rank higher in Google than AWS docs, which are many and scattered. But that is the source of truth.

Shared code was indeed an original use case for why Lambda created layers. Custom runtimes was another. One that I frequently defer to is native modules (like, as Puppeteer, as you mentioned). But layer publishing, permissions, regional considerations, and SAR are (for Architect) unnecessary complications. IMO layers shine for things that don't change super often, or that require some lower-level integration that would be difficult for a regular deployment pipeline. For this reason, I've personally invested quite a few hours into making sure we have [first-class layer support](https://arc.codes/docs/en/reference/configuration/function-config#layers).

For workaday business logic that may be updated frequently and/or that should be easily debuggable, we think our current shared code affordances work well. Vendoring shared code is simple to reason about, deterministic, deploys quickly, is easily debugged, and does not add additional processes to deployments (like layer publishing, layer permission management, etc.). What you see is what you get.

**Reply by aphex** [reply]
Well that makes this an easy one to close then. :) thanks for talking through it.
